src/Compilers/CSharp/Portable/CSharpExtensions.cs

```diff
@@ -200,7 +200,7 @@ public static SyntaxKind Kind(this SyntaxTrivia trivia)
         /// <summary>
         /// Returns <see cref="SyntaxKind"/> for <see cref="SyntaxNode"/> from <see cref="SyntaxNode.RawKind"/> property.
         /// </summary>
-        public static SyntaxKind Kind(this SyntaxNode node)
+        public static SyntaxKind Kind([NotNull] this SyntaxNode node)
```

Comment:
>NotNull [](start = 39, length = 7)

Is this necessary?

Reply:
@tmat yes because lacking this the receiver of the extension method won't be marked as non-null after invocation. Specifically consider this example you asked me to change:

```cs
switch (node.Parent!.Kind()) {
  case ...
    node.Parent.ToString();
}
```

Lacking this annotation the call to `ToString` will be flagged as a potential use of `null`

Reply:
Complete sample:
```cs
using System;
using System.IO;
using System.Threading.Tasks;

#nullable enable
using System.Diagnostics.CodeAnalysis;
class Program{
    static void Main()
    {
        Node? n1 = null;
        n1.M1();        // Warns
        n1.ToString();  // Warns

        Node? n2 = null;
        n2.M2();        // Warns
        n2.ToString();  // Does not warn
    }
}
class Node { }
static class Extensions
{
    internal static void M1(this Node node) { }
    internal static void M2([NotNull] this Node node) { }
}
```

Reply:
I see. It's odd to add such annotation though. Why wouldn't we add `[NotNull]` to all parameters of methods that dereference non-nullable reference typed parameter?

---
In reply to: [362913761](https://github.com/dotnet/roslyn/pull/40699#discussion_r362913761) [](ancestors = 362913761)

Reply:
> Why wouldn't we add [NotNull] to all parameters of methods that dereference non-nullable reference typed parameter?

In this specific case, extension method receiver, it's about expectations. The expectation, as demonstrated by your feedback on the PR and my own head scratching when originally making the changes, is that a call in the form of `a.b()` will ensure that `a` is non-null. I felt it was easier to add this annotation here to reduce confusion in the code bas. 

As to why we don't do it everywhere. It's not wrong to do so just not idiomatic.

Reply:
One more question - in your example above, `M1`'s contract is that `node` is not-null upon entry. Why doesn't the compiler use that to infer that the subsequent dereference of `n1` is ok (after `n1.M1` is called)? Seems like the method shouldn't need to declare the contract twice - once as an input contract and once as an output contract.